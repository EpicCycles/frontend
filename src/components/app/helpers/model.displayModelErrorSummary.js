import {customerAddressFields} from "../model/helpers/fields";
import {displayModelErrorSummary} from "../model/helpers/model";

test('if no error or error_detail undefined returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(undefined);
});
test('if just error no error_detail error returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error"
    };
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe("Model level error");
});
test('if no error with error_detail error_details formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});
test('if error and error_detail formatted returned', () => {
    const model = {
        id: 123,
        address1: "line one",
        address2: "line Ywo",
        address3: "line Three",
        address4: "line Four",
        postcode: "xxxyyy",
        customer: 6,
        error: "Model level error",
        error_detail: {
            "address1": ["A valid address is required.", "SHow me the address is required."],
            "postcode": ["This field may not be blank."]
        },
    };
    const expectedText = "Model level error<br>Address: A valid address is required. SHow me the address is required.<br>Postcode: This field may not be blank.";
    const result = displayModelErrorSummary(model, customerAddressFields);
    expect(result).toBe(expectedText);
});