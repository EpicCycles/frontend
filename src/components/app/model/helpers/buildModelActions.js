/**
 * buildModelActions - builds actions for the current model (use to build additional actions)
 * @param requiredActions - array of objects {iconName, iconTitle, requiresIdOnly, actionFunction}
 * @param model - model being displayed oredited
 * @returns array of model actions specific to current model
 */
export const buildModelActions = (requiredActions, model) => {
  return requiredActions.map(action => {
    const finalAction = {
      iconName: action.iconName,
      iconTitle: action.iconTitle,
    };
    const isSavedModel = model && model.id;
    if (action.requiresIdOnly) {
      finalAction.iconAction = () => isSavedModel && action.actionFunction(model.id);
    } else {
      finalAction.iconAction = () => isSavedModel && action.actionFunction(model);
    }
    finalAction.iconDisabled = !isSavedModel;
    return finalAction;
  });
};
