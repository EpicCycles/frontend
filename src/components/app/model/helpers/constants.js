export const POSTCODE_RULES = [
  { countryCode: 'AT', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'AU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BE', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'BG', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CA', regex: /^[A-Za-z][0-9][A-Za-z] [0-9][A-Za-z][0-9]/, display: 'ANA NAN' },
  { countryCode: 'CH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CY', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'CZ', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'DE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'DK', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'EE', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'ES', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FI', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'FR', regex: /^\d{5}$/, display: 'NNNNN' },
  {
    countryCode: 'GB',
    regex: /^[A-Z](([0-9][0-9]?)|([A-Z][0-9][0-9]?)|([A-Z]?[0-9][A-Z])) ?[0-9][ABD-HJLNP-UW-Z]{2}$/,
    display: 'CCNN NCC',
  },
  { countryCode: 'GR', regex: /^\d{3}\s\d{2}$/, display: 'NNNNN or NNN NN' },
  { countryCode: 'HU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'IT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'KR', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LT', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'LU', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'MT', regex: /^\[A-Z]{3}\s\d{4}$/, display: 'AAA NNNN' },
  { countryCode: 'MY', regex: /^\d{5}$/, display: 'NNNNN' },
  { countryCode: 'NO', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'NZ', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PH', regex: /^\d{4}$/, display: 'NNNN' },
  { countryCode: 'PL', regex: /^\d{2}-\d{3}$/, display: 'NN-NNN' },
  { countryCode: 'PT', regex: /^\d{4}-\d{3}$/, display: 'NNNN-NNN' },
  { countryCode: 'RO', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SE', regex: /^\d{3,4}$/, display: 'NNNNN or NNNN' },
  { countryCode: 'SG', regex: /^\d{6}$/, display: 'NNNNNN' },
  { countryCode: 'SI', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'SK', regex: /^\d{3}\s\d{2}$/, display: 'NNN NN' },
  { countryCode: 'US', regex: /^(\d{5}([-]\d{4})?)$/, display: 'NNNNN-NNNN or NNNNN' },
  { countryCode: 'ZA', regex: /^\d{4}$/, display: 'NNNN' },
];
