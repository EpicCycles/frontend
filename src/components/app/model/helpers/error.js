export const BRAND_NAME_MISSING = "A name is required for the brand";
export const FRAME_NAME_MISSING = "A name is required for the frame";
export const ATTRIBUTE_NAME_MISSING = "A name is required for the attribute";
export const BUNDLE_NAME_MISSING = "A name is required for the bundle";
export const MODEL_NAME_MISSING = "A name is required for the model";
export const PART_NAME_MISSING = "A name is required for the part";
export const PART_TYPE_NAME_MISSING = "A name is required for the part type";
export const SECTION_MISSING = "A section must be selected for the part type";
export const PART_TYPE_MISSING = "A part type must be selected";
export const BRAND_MISSING = "A brand must be selected";
export const PRODUCTS_MISSING = "At least 2 products must be selected";
export const SUPPLIER_MISSING = "A supplier must be selected";
export const PART_MISSING = "A part must be selected";
export const ADDRESS_MISSING = "At least one line of address must be entered";
export const COUNTRY_MISSING = "A country must be selected";
export const INVALID_POSTCODE = "A postcode is expected which for this country has the format ";
export const SELECT_ONE_MISSING = "Please select an option.";
export const VALUE_MISSING = "A value is required.";
export const INVALID_EMAIL = "Invalid email address.";
export const INVALID_URL = "Invalid URL";
