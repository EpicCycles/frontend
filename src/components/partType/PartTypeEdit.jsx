import React, {Fragment} from "react";
import {generateRandomCode, updateObject} from "../../helpers/utils";
import {Icon} from "semantic-ui-react";
import {NEW_ELEMENT_ID} from "../../helpers/constants";
import PartTypeData from "./PartTypeData";
import {processPartTypeValueChanges} from "./helpers/partType";

class PartTypeEdit extends React.Component {
    handlePartTypeValueChange = (fieldName, input) => {
        const updatedPartType = processPartTypeValueChanges(this.props.partType, this.props.componentKey,fieldName, input);

        this.props.updatePartType(this.props.componentKey, updatedPartType);
    };

    handleInputClear = (fieldName) => {
        if (window.confirm("Please confirm that you want to delete this Part Type")) {
            const updatedPartType = updateObject(this.props.partType);
            updatedPartType.delete = true;
            this.props.updatePartType(this.props.componentKey, updatedPartType);
        }
    };
    toggleDetail = () => {
        this.handlePartTypeValueChange(`detail_${this.props.componentKey}`, !this.props.partType._detail)
    };
    addAnother = () => {
        const updatedPartType = updateObject(this.props.partType);
        updatedPartType.dummyKey = generateRandomCode();
        this.props.updatePartType(NEW_ELEMENT_ID, updatedPartType);
    };

    render() {
        const { partType, componentKey, updatePartType } = this.props;
        return <Fragment>
            <td>
                {componentKey !== NEW_ELEMENT_ID ?
                    <Icon
                        name={`toggle ${partType._detail ? "down" : "right"}`}
                        onClick={this.toggleDetail}
                    />
                    :
                    <Icon
                        name="add"
                        onClick={this.addAnother}
                    />
                }
            </td>
            <td>
                <PartTypeData
                    partType={partType}
                    componentKey={componentKey}
                    handlePartTypeValueChange={this.handlePartTypeValueChange}
                    updatePartType={updatePartType}
                    />
            </td>
        </Fragment>;
    }
}

export default PartTypeEdit;