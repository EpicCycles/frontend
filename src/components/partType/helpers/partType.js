import { updateObject } from '../../../helpers/utils';
import { PART_TYPE_NAME_MISSING } from '../../app/model/helpers/error';
import { NEW_ELEMENT_ID } from '../../../helpers/constants';
import { getModelKey } from '../../app/model/helpers/model';

export const getPartTypeForName = (sections, partTypeName) => {
  let foundPartType;
  sections.some(section => {
    foundPartType = section.partTypes.find(partType =>
      doesFieldMatchPartType(partType, partTypeName),
    );
    return !!foundPartType;
  });
  return foundPartType;
};
export const getPartTypeOptions = sections => {
  const partTypeOptions = [];
  sections &&
    sections.forEach(section => {
      section.partTypes.forEach(partType => {
        partTypeOptions.push({
          value: String(getModelKey(partType)),
          name: partType.name + ' (' + section.name + ')',
        });
      });
    });
  return partTypeOptions;
};
export const getPartTypeName = (partTypeId, sections) => {
  if (!partTypeId) return undefined;
  const partType = getPartType(partTypeId, sections);
  if (partType) return partType.name;
  return 'Unknown Part Type';
};
export const getPartType = (partTypeToFind, sections) => {
  if (Number.isNaN(partTypeToFind)) return undefined;
  const partTypeId = Number(partTypeToFind);
  let partType;
  sections.some(section => {
    return section.partTypes.some(aPartType => {
      if (aPartType.id === partTypeId) {
        partType = aPartType;
        return true;
      }
      return false;
    });
  });
  return partType;
};
export const getPartTypeAttributeFields = (partTypeId, sections) => {
  const partType = getPartType(partTypeId, sections);
  if (partType) return partType.attributes;
  return [];
};
export const processPartTypeValueChanges = (partType, componentKey, fieldName, input) => {
  const updatedPartType = updateObject(partType);
  if (fieldName.startsWith('name')) updatedPartType.name = input;
  if (!updatedPartType.name) {
    updatedPartType.error = true;
    updatedPartType.error_detail = PART_TYPE_NAME_MISSING;
  } else {
    updatedPartType.error = false;
    updatedPartType.error_detail = '';
  }
  if (fieldName.startsWith('description')) updatedPartType.description = input;
  if (fieldName.startsWith('can_be_substituted')) updatedPartType.can_be_substituted = input;
  if (fieldName.startsWith('can_be_omitted')) updatedPartType.can_be_omitted = input;
  if (fieldName.startsWith('customer_visible')) updatedPartType.customer_visible = input;
  if (fieldName.startsWith('attributes')) updatedPartType.attributes = input;
  if (fieldName.startsWith('synonyms')) updatedPartType.synonyms = input;
  if (fieldName.startsWith('detail')) updatedPartType._detail = input;
  if (componentKey === NEW_ELEMENT_ID) updatedPartType.dummyKey = NEW_ELEMENT_ID;

  updatedPartType.changed = true;
  return updatedPartType;
};
export const doesFieldMatchPartType = (partType, fieldName) => {
  const fieldNameLower = fieldName.toLowerCase();
  if (partType.name.toLowerCase() === fieldNameLower) {
    return true;
  } else {
    return partType.synonyms.some(synonym => {
      return synonym.name.toLowerCase() === fieldNameLower;
    });
  }
};
export const NEW_PART_TYPE = {
  attributes: [],
  name: '',
  can_be_substituted: true,
  can_be_omitted: true,
  customer_visible: true,
};
export const NEW_ATTRIBUTE = {
  attribute_name: '',
  in_use: true,
  mandatory: false,
  attribute_type: 1,
};
export const attributeSummary = attribute => {
  let attributeDetail = [attribute.attribute_name];
  if (attribute.in_use) attributeDetail.push(' in use');
  if (attribute.mandatory) attributeDetail.push(' must be entered');
  if (attribute.options && attribute.options.length > 0) {
    attributeDetail.push(
      ` allowed options: ${attribute.options
        .map(option => {
          return option.option_name;
        })
        .toString()}`,
    );
  }
  return attributeDetail.toString();
};
export const attributePlaceholder = partType => {
  let placeholder = [];
  if (partType && partType.attributes)
    partType.attributes.forEach(attribute => {
      if (attribute.in_use) {
        if (attribute.options && attribute.options.length > 0) {
          placeholder.push(
            `${attribute.attribute_name}(${attribute.options
              .map(option => {
                return option.option_name;
              })
              .toString()})`,
          );
        } else {
          placeholder.push(attribute.attribute_name);
        }
      }
    });
  return placeholder.join(', ');
};
