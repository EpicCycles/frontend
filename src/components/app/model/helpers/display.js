import {
  BIKE,
  BRAND,
  CHECKBOX,
  CURRENCY,
  CUSTOMER,
  DATE_TIME,
  NUMBER,
  PART_TYPE,
  RADIO,
  SELECT_ONE,
  SUPPLIER,
  USER,
} from './fields';
import { getCustomerName } from '../../../customer/helpers/customer';
import { getBrandName } from '../../../brand/helpers/brand';
import { getNameForValue } from './model';
import { getSupplierName } from '../../../supplier/helpers/supplier';
import { getPartTypeName } from '../../../partType/helpers/partType';
import { getBikeName } from '../../../bike/helpers/bike';
import { getUserName } from '../../../user/helpers/user';
import { CHARGE } from '../../../quoteCharge/helpers/quoteChargeFields';
import { chargeName } from '../../../charge/helpers/chargeName';
import { findObjectWithKey } from '../../../../helpers/utils';
import { fittingText } from '../../../fitting/helpers/fitting';
import { FITTING } from '../../../quote/helpers/quoteFields';

export const fixedHeaderClassname = lockColumn => {
  if (lockColumn) return 'grid-header--fixed-left';
  return '';
};
export const fixedDetailsClassname = lockColumn => {
  if (lockColumn) return 'grid-item--fixed-left';
  return '';
};
export const gridHeaderClass = (baseClassName = '', fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && fieldIndex === 0;
  return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = '', fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && fieldIndex === 0;
  return `grid-item ${fixedDetailsClassname(shouldLock)} ${baseClassName}`;
};
export const formattedDate = date => {
  if (date) return date.toLocaleDateString('en-GB');
  return '';
};
export const formattedDateTime = date => {
  if (date) return date.toLocaleString('en-GB');
  return '';
};
export const formattedPrice = fieldValue => {
  if (fieldValue)
    return Number(fieldValue).toLocaleString('en-GB', {
      style: 'currency',
      currency: 'GBP',
    });
  return '';
};
export const fieldAlignment = field => {
  switch (field.type) {
    case NUMBER:
    case CURRENCY:
      return 'align_right';
    case CHECKBOX:
    case RADIO:
      return 'align_center';
    default:
      return '';
  }
};
export const buildViewString = (model, field, sourceDataArrays = {}) => {
  let viewData;
  const {
    sections,
    brands,
    suppliers,
    customers,
    bikes,
    frames,
    users,
    charges,
    fittings,
  } = sourceDataArrays;
  const fieldValue = model ? model[field.fieldName] : undefined;
  switch (field.type) {
    case CUSTOMER:
      viewData = fieldValue ? getCustomerName(fieldValue, customers) : '';
      break;
    case CURRENCY:
      viewData = formattedPrice(fieldValue);
      break;
    case CHECKBOX:
      viewData = fieldValue ? 'Yes' : 'No';
      break;
    case DATE_TIME:
      viewData = fieldValue ? formattedDateTime(new Date(fieldValue)) : '';
      break;
    case PART_TYPE:
      viewData = fieldValue ? getPartTypeName(fieldValue, sections) : '';
      break;
    case BRAND:
      viewData = fieldValue ? getBrandName(fieldValue, brands) : '';
      break;
    case SUPPLIER:
      viewData = fieldValue ? getSupplierName(fieldValue, suppliers) : '';
      break;
    case BIKE:
      viewData = fieldValue ? getBikeName(fieldValue, bikes, frames, brands) : '';
      break;
    case SELECT_ONE:
      viewData = fieldValue ? getNameForValue(fieldValue, field.selectList) : '';
      break;
    case USER:
      viewData = fieldValue ? getUserName(fieldValue, users) : '';
      break;
    case CHARGE:
      viewData = fieldValue ? chargeName(fieldValue, charges) : '';
      break;
    case FITTING:
      const fitting = fieldValue ? findObjectWithKey(fittings, fieldValue) : undefined;
      viewData = fitting ? fittingText(fitting) : '';
      break;
    default:
      viewData = fieldValue ? fieldValue : '';
  }
  return viewData;
};
