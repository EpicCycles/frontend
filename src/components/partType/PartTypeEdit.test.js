import PartTypeEdit from './PartTypeEdit';
import { NEW_ELEMENT_ID } from '../../helpers/constants';

test('displays the PartTypeEdit component for an existing part type with attributes hidden', () => {
  const componentKey = '21';
  const partType = {
    _detail: false,
    id: 21,
    name: 'part one',
    can_be_substituted: false,
    can_be_omitted: true,
    customer_visible: true,
  };
  const component = shallow(<PartTypeEdit componentKey={componentKey} partType={partType} />);
  expect(toJson(component)).toMatchSnapshot();
});
test('displays the PartTypeEdit component for an existing part type with attributes', () => {
  const componentKey = 'dummy2';
  const partType = {
    _detail: true,
    dummyKey: componentKey,
    name: 'part one',
    can_be_substituted: false,
    can_be_omitted: true,
    customer_visible: true,
    attributes: ['attribute1', 'attribute2'],
  };
  const component = shallow(<PartTypeEdit componentKey={componentKey} partType={partType} />);
  expect(toJson(component)).toMatchSnapshot();
});
test('displays the PartTypeEdit component for a new part type with no attributes', () => {
  const partType = {
    _detail: true,
    dummyKey: NEW_ELEMENT_ID,
    name: 'part one',
    can_be_substituted: false,
    can_be_omitted: true,
    customer_visible: true,
  };
  const component = shallow(<PartTypeEdit componentKey={NEW_ELEMENT_ID} partType={partType} />);
  expect(toJson(component)).toMatchSnapshot();
});
