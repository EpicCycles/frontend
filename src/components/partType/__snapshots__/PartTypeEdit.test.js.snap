// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the PartTypeEdit component for a new part type with no attributes 1`] = `
<Fragment>
  <td>
    <AddLink
      addFunction={[Function]}
      addObjectName="Type"
    />
  </td>
  <td>
    <PartTypeData
      componentKey="new"
      handlePartTypeValueChange={[Function]}
      partType={
        Object {
          "_detail": true,
          "can_be_omitted": true,
          "can_be_substituted": false,
          "customer_visible": true,
          "dummyKey": "new",
          "name": "part one",
        }
      }
    />
  </td>
</Fragment>
`;

exports[`displays the PartTypeEdit component for an existing part type with attributes 1`] = `
<Fragment>
  <td>
    <Icon
      as="i"
      name="toggle down"
      onClick={[Function]}
    />
  </td>
  <td>
    <PartTypeData
      componentKey="dummy2"
      handlePartTypeValueChange={[Function]}
      partType={
        Object {
          "_detail": true,
          "attributes": Array [
            "attribute1",
            "attribute2",
          ],
          "can_be_omitted": true,
          "can_be_substituted": false,
          "customer_visible": true,
          "dummyKey": "dummy2",
          "name": "part one",
        }
      }
    />
  </td>
</Fragment>
`;

exports[`displays the PartTypeEdit component for an existing part type with attributes hidden 1`] = `
<Fragment>
  <td>
    <Icon
      as="i"
      name="toggle right"
      onClick={[Function]}
    />
  </td>
  <td>
    <PartTypeData
      componentKey="21"
      handlePartTypeValueChange={[Function]}
      partType={
        Object {
          "_detail": false,
          "can_be_omitted": true,
          "can_be_substituted": false,
          "customer_visible": true,
          "id": 21,
          "name": "part one",
        }
      }
    />
  </td>
</Fragment>
`;
